File: server/app/api/mcp.py
"""MCP API：服务管理 + 本平台对外暴露的 MCP 端点。"""

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session as DbSession

from app.db.database import get_db
from app.mcp import service as svc

router = APIRouter(prefix="/api/mcp", tags=["mcp"])


class ServerIn(BaseModel):
    name: str
    url: str
    auth_header: str = ""
    require_approval: bool = False


class ServerPatch(BaseModel):
    name: str | None = None
    url: str | None = None
    auth_header: str | None = None
    enabled: bool | None = None
    require_approval: bool | None = None


@router.get("/servers")
def list_servers(db: DbSession = Depends(get_db)):
    return svc.list_servers(db)


@router.post("/servers")
def add_server(body: ServerIn, db: DbSession = Depends(get_db)):
    result = svc.add_server(db, body.name, body.url, body.auth_header, body.require_approval)
    if result.get("error"):
        raise HTTPException(400, result["error"])
    return result


@router.patch("/servers/{server_id}")
def update_server(server_id: int, body: ServerPatch, db: DbSession = Depends(get_db)):
    result = svc.update_server(db, server_id, **body.model_dump(exclude_none=True))
    if result.get("error"):
        raise HTTPException(404, result["error"])
    return result


@router.delete("/servers/{server_id}")
def delete_server(server_id: int, db: DbSession = Depends(get_db)):
    result = svc.delete_server(db, server_id)
    if result.get("error"):
        raise HTTPException(404, result["error"])
    return result


@router.post("/servers/{server_id}/test")
def test_server(server_id: int, db: DbSession = Depends(get_db)):
    return svc.test_server(db, server_id)


@router.post("/servers/{server_id}/sync")
def sync_server(server_id: int, db: DbSession = Depends(get_db)):
    return svc.sync_tools(db, server_id)


# ---------------------------------------------------------------------------
# 本平台对外暴露的 MCP 服务（JSON-RPC 2.0）
# ---------------------------------------------------------------------------

class RpcIn(BaseModel):
    jsonrpc: str = "2.0"
    id: int | str | None = 1
    method: str
    params: dict | None = None


@router.post("")
def mcp_endpoint(body: RpcIn, db: DbSession = Depends(get_db)):
    """让其他 Agent / 客户端把本平台当作 MCP 工具服务器使用。"""
    from app.agent.tools import build_registry

    registry = build_registry()

    def ok(result: dict):
        return {"jsonrpc": "2.0", "id": body.id, "result": result}

    if body.method == "initialize":
        return ok({
            "protocolVersion": "2025-06-18",
            "capabilities": {"tools": {}},
            "serverInfo": {"name": "risk-warning-platform", "version": "0.3"},
        })
    if body.method == "ping":
        return ok({})
    if body.method == "tools/list":
        tools = []
        for schema in registry.definitions():
            fn = schema["function"]
            tools.append({
                "name": fn["name"],
                "description": fn["description"],
                "inputSchema": fn.get("parameters") or {"type": "object", "properties": {}},
            })
        return ok({"tools": tools})
    if body.method == "tools/call":
        params = body.params or {}
        name = params.get("name")
        args = params.get("arguments") or {}
        if not name:
            return {"jsonrpc": "2.0", "id": body.id, "error": {"code": -32602, "message": "缺少工具名"}}
        result = registry.call(name, args, db)
        return ok({"content": [{"type": "text", "text": _as_text(result)}]})
    return {"jsonrpc": "2.0", "id": body.id, "error": {"code": -32601, "message": f"未知方法: {body.method}"}}


def _as_text(result) -> str:
    import json

    if isinstance(result, str):
        return result
    return json.dumps(result, ensure_ascii=False)

File: backend/src/api/mcp.rs
//! MCP 服务管理 API：`/api/mcp/*`。

use axum::extract::{Path, State};
use axum::Json;
use serde::Deserialize;
use serde_json::{json, Value};

use crate::error::{AppError, AppResult};
use crate::services::mcp;
use crate::state::AppState;

/// GET /api/mcp/servers
pub async fn list(State(st): State<AppState>) -> AppResult<Json<Value>> {
    Ok(Json(mcp::list_servers(&st.db)?))
}

#[derive(Deserialize)]
pub struct ServerIn {
    pub name: String,
    pub url: String,
    #[serde(default)]
    pub auth_header: String,
    #[serde(default)]
    pub require_approval: bool,
}

/// POST /api/mcp/servers —— 注册并立即同步工具
pub async fn create(
    State(st): State<AppState>,
    Json(body): Json<ServerIn>,
) -> AppResult<Json<Value>> {
    let result =
        mcp::add_server(&st.db, &body.name, &body.url, &body.auth_header, body.require_approval).await?;
    if let Some(err) = result.get("error").and_then(|v| v.as_str()) {
        return Err(AppError::bad_request(err));
    }
    Ok(Json(result))
}

/// PATCH /api/mcp/servers/{id}
pub async fn update(
    State(st): State<AppState>,
    Path(id): Path<i64>,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let result = mcp::update_server(&st.db, id, &body)?;
    if let Some(err) = result.get("error").and_then(|v| v.as_str()) {
        return Err(AppError::not_found(err));
    }
    Ok(Json(result))
}

/// DELETE /api/mcp/servers/{id}
pub async fn delete(State(st): State<AppState>, Path(id): Path<i64>) -> AppResult<Json<Value>> {
    let result = mcp::delete_server(&st.db, id)?;
    if let Some(err) = result.get("error").and_then(|v| v.as_str()) {
        return Err(AppError::not_found(err));
    }
    Ok(Json(result))
}

/// POST /api/mcp/servers/{id}/test
pub async fn test_server(State(st): State<AppState>, Path(id): Path<i64>) -> AppResult<Json<Value>> {
    let result = mcp::test_server(&st.db, id).await?;
    if let Some(err) = result.get("error").and_then(|v| v.as_str()) {
        return Err(AppError::not_found(err));
    }
    Ok(Json(result))
}

/// POST /api/mcp/servers/{id}/sync
pub async fn sync_server(State(st): State<AppState>, Path(id): Path<i64>) -> AppResult<Json<Value>> {
    let result = mcp::sync_tools(&st.db, id).await?;
    if let Some(err) = result.get("error").and_then(|v| v.as_str()) {
        return Err(AppError::not_found(err));
    }
    Ok(Json(result))
}

/// POST /api/mcp —— **服务端**：让其他 Agent / 客户端把本平台当作 MCP 工具服务器使用。
///
/// 支持 initialize / ping / tools/list / tools/call（与 Python 版 `app/api/mcp.py` 对齐）。
#[derive(Deserialize)]
pub struct RpcIn {
    #[serde(default)]
    pub id: Option<Value>,
    pub method: String,
    #[serde(default)]
    pub params: Option<Value>,
}

fn ok(id: &Option<Value>, result: Value) -> Json<Value> {
    Json(json!({
        "jsonrpc": "2.0",
        "id": id.clone().unwrap_or(json!(1)),
        "result": result,
    }))
}

fn rpc_error(id: &Option<Value>, code: i64, message: String) -> Json<Value> {
    Json(json!({
        "jsonrpc": "2.0",
        "id": id.clone().unwrap_or(json!(1)),
        "error": { "code": code, "message": message },
    }))
}

pub async fn rpc_endpoint(
    State(st): State<AppState>,
    Json(body): Json<RpcIn>,
) -> AppResult<Json<Value>> {
    let registry = crate::agent::tools::build_registry();
    match body.method.as_str() {
        "initialize" => Ok(ok(
            &body.id,
            json!({
                "protocolVersion": "2025-06-18",
                "capabilities": { "tools": {} },
                "serverInfo": { "name": "risk-warning-platform", "version": env!("CARGO_PKG_VERSION") },
            }),
        )),
        "ping" => Ok(ok(&body.id, json!({}))),
        "tools/list" => {
            let tools: Vec<Value> = registry
                .definitions()
                .iter()
                .filter_map(|schema| {
                    let f = schema.get("function")?;
                    Some(json!({
                        "name": f.get("name")?,
                        "description": f.get("description").cloned().unwrap_or(json!("")),
                        "inputSchema": f
                            .get("parameters")
                            .cloned()
                            .unwrap_or_else(|| json!({ "type": "object", "properties": {} })),
                    }))
                })
                .collect();
            Ok(ok(&body.id, json!({ "tools": tools })))
        }
        "tools/call" => {
            let params = body.params.clone().unwrap_or(json!({}));
            let Some(name) = params.get("name").and_then(|v| v.as_str()) else {
                return Ok(rpc_error(&body.id, -32602, "缺少工具名".into()));
            };
            let args = params.get("arguments").cloned().unwrap_or(json!({}));
            let result = if crate::agent::tools::is_async_tool(name) {
                crate::agent::tools::call_tool_async(&st, name, &args).await
            } else {
                crate::agent::tools::call_tool(&st.db, name, &args, &[])
            };
            let text = if let Some(s) = result.as_str() {
                s.to_string()
            } else {
                serde_json::to_string(&result).unwrap_or_else(|_| "{}".into())
            };
            Ok(ok(&body.id, json!({ "content": [{ "type": "text", "text": text }] })))
        }
        other => Ok(rpc_error(&body.id, -32601, format!("未知方法: {other}"))),
    }
}
